spec/text-editor-spec.coffee

```diff
@@ -2187,12 +2187,37 @@ describe "TextEditor", ->
         expect(editor.lineTextForBufferRow(1)).toBe '  '
         expect(editor.lineTextForBufferRow(2)).toBe '}'
 
-    describe "when a new line is appended before a closing tag (e.g. by pressing enter before a selection)", ->
-      it "moves the line down and keeps the indentation level the same when editor.autoIndent is true", ->
-        atom.config.set('editor.autoIndent', true)
-        editor.setCursorBufferPosition([9,2])
-        editor.insertNewline()
-        expect(editor.lineTextForBufferRow(10)).toBe '  };'
+    describe ".insertNewLine()", ->
+      describe "when a new line is appended before a closing tag (e.g. by pressing enter before a selection)", ->
+        it "moves the line down and keeps the indentation level the same when editor.autoIndent is true", ->
+          atom.config.set('editor.autoIndent', true)
+          editor.setCursorBufferPosition([9,2])
+          editor.insertNewline()
+          expect(editor.lineTextForBufferRow(10)).toBe '  };'
+
+      describe "when a newline is appended with a trailing closing tag behind the cursor (e.g. by pressing enter in the middel of a line)", ->
+        it "indents the new line to the correct level when editor.autoIndent is true and using a curly-bracket language", ->
+          atom.config.set('editor.autoIndent', true)
+          editor.setText('var test = function () {\n  return true;};')
+          editor.setCursorBufferPosition([1,14])
+          editor.insertNewline()
+          expect(editor.lineTextForBufferRow(1)).toBe '  return true;'
+          expect(editor.lineTextForBufferRow(2)).toBe '};'
+
+        it "indents the new line to the correct level when editor.autoIndent is true and using a off-side rule language", ->
+          waitsForPromise ->
+            atom.workspace.open(null, autoIndent: true).then (o) -> editor = o
+
+          waitsForPromise ->
+            atom.packages.activatePackage('language-coffee-script')
+
+          runs ->
+            editor.setText('if true\n  return trueelse\n  return false')
+            editor.setCursorBufferPosition([1,13])
+            editor.insertNewline()
+            expect(editor.lineTextForBufferRow(1)).toBe '  return true'
+            expect(editor.lineTextForBufferRow(2)).toBe 'else'
+            expect(editor.lineTextForBufferRow(3)).toBe '  return false'
```

Comment:
Yes, you are absolutely right! Please see the first comment I made to this PR, as the existing tests are to verify that behaviour. And indeed they do not require any changes to the existing regexps as they seem to fix something that wasn't working as expected before.

As for the behaviour as described in the 3rd comment, for that indeed a regexp should be added to a language that needs/uses such behaviour. Go does, so I created a PR there which was merged just a few hours ago (https://github.com/atom/language-go/pull/44) whit an additional regexp. So now that this one is merged I can add some more tests to verify the additional behaviour as well. Will do that shortly...